cmd/kubeadm/app/util/version.go

```diff
@@ -92,17 +93,24 @@ func KubernetesReleaseVersion(version string) (string, error) {
 			if body != "" {
 				return "", err
 			}
-			// Handle air-gapped environments by falling back to the client version.
-			klog.Infof("could not fetch a Kubernetes version from the internet: %v", err)
-			klog.Infof("falling back to the local client version: %s", clientVersion)
-			return KubernetesReleaseVersion(clientVersion)
+			if clientVersionErr == nil {
+				// Handle air-gapped environments by falling back to the client version.
+				klog.Infof("could not fetch a Kubernetes version from the internet: %v", err)
+				klog.Infof("falling back to the local client version: %s", clientVersion)
+				return KubernetesReleaseVersion(clientVersion)
+			}
 		}
 
-		if clientVersionErr != nil {
+		if err == nil && clientVersionErr != nil {
 			klog.Warningf("could not obtain client version; using remote version: %s", body)
 			return KubernetesReleaseVersion(body)
 		}
 
+		if err != nil && clientVersionErr != nil {
+			klog.Warningf("could not obtain neither client nor remote version; fall back to: %s", constants.CurrentKubernetesVersion)
+			return KubernetesReleaseVersion(constants.CurrentKubernetesVersion.String())
```

Comment:
is there a unit test to ensure this constant is always correct? if not, it is very likely to drift and be forgotten

Reply:
> these should be the only constants that we bump on each minor release and we must bump them anyway

Doesn't the current release constant require bumping on every patch version as well?

Reply:
That seems like a misleading constant, then, and one that would be likely to be used incorrectly.

Reply:
Honestly speaking this recursive function with side effects smells suboptimal to me since it clearly violates the Single Responsibility principle. I would rather try to constrain ClusterConfig.KubernetesVersion to be a canonical version only and always. Then at config parsing/cmdline option setting time we'd detect whether the provided value is

1. a version or
2. just a version label supposed to be resolved remotely into a version or
3. nil.

In the first case `normalizedBuildVersion()` would be used once. In the second case `KubernetesReleaseVersion()` would be limited to resolving the remote version only. And in the third case we could try to read the `gitVersion` and fail if it's not set.

Then the only authoritative source of the version value would be `ClusterConfig.KubernetesVersion`. With this approach the unit tests would not touch networking.

Reply:
> Honestly speaking this recursive function with side effects smells suboptimal to me since it clearly violates the Single Responsibility principle

i support your concern here and the recursion can be avoided.
@kad and @luxas probably have opinions here as they authored the original?

>  I would rather try to constrain ClusterConfig.KubernetesVersion to be a canonical version only and always

isn't this the case already? if the user does not feed a KubernetesVersion and a config, the version has to come from somewhere. internet endpoint, client version, constant?

> With this approach the unit tests would not touch networking.

there are calls like this one:
https://github.com/kubernetes/kubernetes/blob/b502b99d25fe8297306bfab93eae052e0a852354/cmd/kubeadm/app/cmd/token.go#L213-L215

that pre-set a version so that the version is later not fetched from the internet for commands that don't need internet. but this still needs a client that was built properly.

so potentially, unit tests can do the same.

Reply:
> isn't this the case already? if the user does not feed a KubernetesVersion and a config, the version has to come from somewhere. internet endpoint, client version, constant?

IIRC some unit tests don't (can't) consult `ClusterConfig.KubernetesVersion` relying solely on `KubernetesReleaseVersion()`. I'll try to come up with a proposal.

Reply:
... in a separate PR that is.